### ISSUE TITLE
Inclusion in the Coq Platform

### ISSUE THREAD

**Issue by MSoegtropIMC** [lukaszcz/coqhammer#110]
Dear Coq-Hammer Team,

@palmskog requested here: (https://github.com/coq/platform/issues/10) to include your development into the Coq Platform, which is the standard user facing distribution of Coq.

Coq Platform has a "full" and "extended" level. Inclusion in the "full" level requires an explicit statement by the maintainers, that they agree to the [charter](https://github.com/coq/platform/blob/2021.02/charter.md) of the Coq Platform and intend to publish compatible releases for each release of Coq in a reasonable time frame. For the "extended" level the rules are more relaxed. For developments in the "full" level, users should be able to rely on that the package is maintained, so that they can base their own development on it without a large risk that they have to factor it out again later cause of maintenance issues.

Can you please comment in the Coq Platform issue mentioned above if you do want your package included in the Coq Platform in agreement to the Coq Platform [charter](https://github.com/coq/platform/blob/2021.02/charter.md), and if so which level you would prefer?

Best regards,

Michael

**Comment by lukaszcz** [comment]
CoqHammer is now fairly stable and designed in such a way that keeping it
up-to-date with changes in Coq sources usually doesn't require many
adaptations. Creating a new release is thus usually easy, and I can do it
quickly. However, I cannot commit to maintaining a working Windows version.
I can only commit to keeping it working and up-to-date on Mac and Linux.
Someone else would need to see about a Windows version.

**Comment by palmskog** [comment]
@MSoegtropIMC to give some additional context on the project structure and potential issues with Windows, the project has two components, represented as opam packages built from the same release archive:

1. The general proof search tactics (`sauto` and friends, package `coq-hammer-tactics`), which to my knowledge are implemented as a regular Coq plugin in OCaml complemented by some Ltac. I expect this package to work without any changes on Windows (the package can be built both with coq_makefile and Dune).
2. The Hammer part (package `coq-hammer`), which is a combination of (a) a Coq plugin in OCaml that uses `sauto` and other tactics above, (b) two project-hosted CLI tools in C/C++ compiled during package installation, and (c) CLI tools provided by ATPs. To my knowledge, this is where the Unix-specific assumptions reside, which may or may not work with a Coq platform cygwin environment.

**Comment by lukaszcz** [comment]
One more potential problem is that the OCaml code uses some functions from
the Unix module that are "not implemented" on Windows (like "fork" etc for
implementing the timeout mechanism). But I presume this should work in the
cygwin environment which is supposed to be POSIX-compatible.

I just don't have a personal computer with Windows installed and I don't
intend to buy Windows. So I can't test this out. Well, maybe once I could,
but not with every new Coq release.

**Comment by lukaszcz** [comment]
One more thing: you need to have "grep" in the path to use the "hammer"
tactic. It was just the easiest way to do output filtering when invoking
the provers. Probably this should be replaced by using some OCaml library,
but as long as it works (on Max and Linux), I had no motivation to do it.

**Comment by MSoegtropIMC** [comment]
Thanks for the detailed information. I will do a Windows build today and see what happens ...

For testing the machinery, is it sufficient to have just eprover (should be trivial to compile with the opam package I did) or do different ATPs different levels of massaging input and output?

**Comment by lukaszcz** [comment]
With only one prover you just may get a lower success rate because eprover
might not be able to handle every example. The system is modular - the
calling of the provers done in the last step and is entirely independent of
everything else.

**Comment by MSoegtropIMC** [comment]
@lukaszcz : the plan is to have eprover and z3 for a start - booth should be easy. We can look into opam packages for the others later. For z3 I plan to make an opam package for the TPTP variant, which depends on the main z3 package.

One question: do I need the provers during compile time or during run time? I do have eprover in the path but the hammer tactic says it didn't find any provers.

**Comment by MSoegtropIMC** [comment]
Regarding Windows: I would say the deal is that the Coq Platform creates an environment which is sufficiently equal on all supported platforms, so that one can expect that if it works on one, it does work on another.

The only common issue is comparing test results against reference files - these contain CR on Windows, but it is easy to get around this. Also your parsers of the ATP output (if it si text) should accept CRs. Other deviations are pretty rare - e.g. the gappa plugin, which also calls an external prover - did work out of the box.

**Comment by MSoegtropIMC** [comment]
I had a look at Windows. There are a few issues:

1.) you frequently redirect to /dev/nul - this does not work on Windows because there is no file /dev/nul (otherwise redirection does work and has the samy syntax, just the file name is different - on Windows it would be NULL)

2.) During make install the two helper exes `htimeout` and `predict` don't end up where they should - a backslash quoting issue.

3.) After fixing this the provers are detected and I think called, but at some point you do a real Unix fork - the one which implements page sharing between the forks. Do you actually need this or might some simpler form of subprocess createion suffice?

**Comment by lukaszcz** [comment]
>
>
> One question: do I need the provers during compile time or during run
> time? I do have eprover in the path but the hammer tactic says it didn't
> find any provers.
>
>
> The provers are needed only during runtime. They must have exactly the
names described in CoqHammer installation instructions.

**Comment by lukaszcz** [comment]
> 1.) you frequently redirect to /dev/nul - this does not work on Windows
> because there is no file /dev/nul (otherwise redirection does work and has
> the samy syntax, just the file name is different - on Windows it would be
> NULL)
>
This should be easy to fix.

> 2.) During make install the two helper exes htimeout and predict don't
> end up where they should - a backslash quoting issue.
>
Easy fix.

> 3.) After fixing this the provers are detected and I think called, but at
> some point you do a real Unix fork - the one which implements page sharing
> between the forks. Do you actually need this or might some simpler form of
> subprocess createion suffice?
>
Not sure, I'm not entirely familiar with how subprocess creation on Windows
works. The child processes need the memory contents from the parent process
at the time of creation, but then they do not communicate through
shared memory with the parent or each other. They communicate via exit
status and unix pipes (I think Unix.pipe works on Windows, at least there's
no mention in OCaml documentation that it doesn't).

**Comment by lukaszcz** [comment]
The fork is used for parallel invocations. If you disable parallelism
then CoqHammer doesn't use fork -- the provers are ultimately invoked
with Sys.command which should work out of the box. So the broader
question is how to correctly implement parallel external command and
internal tactic executions on Windows.

**Comment by lukaszcz** [comment]
There are three places where fork is used:

1. src/lib/hhpartac.ml
This implements parallel tactic invocation.

2. src/plugin/parallel.ml
This implements parallel invocation of OCaml functions.

3. src/plugin/timeout.ml
This implements a timeout mechanism. Maybe now this can be replaced by
Coq's timeout mechanism, but at the time I wrote it Coq's timeout was
too buggy.

One can ignore the forks in hammer_main.ml - they are only used for
batch evaluations and will never be called in interactive use.

**Comment by MSoegtropIMC** [comment]
@lukaszcz : can you tell me how to disable parallelism? I can test this, but from my (very limited) understanding of how this works I would guess that parallelism is somehow required, since different provers could use substantially different time to solve a goal.

In general process creation in Windows does not inherit memory in any way - one creates a new process from scratch. A child process can inherit e.g. handles, though - details depend on the way processes are created and I am not an expert of the OCaml wrappers. For parallelism it is more common to create threads (which obviously share memory) and then possibly create processes from the threads.

The timeout should be easiest to fix on Windows - worst case one can create separate code which has the same functionality and interface.

I would have to have a deeper look into the parallel execution. I guess the goal is to spawn several ATP processes in parallel in a controlled way. I would think the easiest portable way of doing this would be using the OCaml thread library and spawning processes from the threads.

**Comment by lukaszcz** [comment]
>
>
> @lukaszcz <https://github.com/lukaszcz> : can you tell me how to disable
> parallelism? I can test this, but from my (very limited) understanding of
> how this works I would guess that parallelism is somehow required, since
> different provers could use substantially different time to solve a goal.
>

Set Hammer GSMode 0.
Unset Hammer Parallel.

This disables parallelism for the provers only. There's no way for the user
to disable parallelism for the reconstruction tactics, or to use a non-fork
timeout.

But parallelism is a major feature -- trying different provers and
reconstruction tactics sequentially just takes too long.


> The timeout should be easiest to fix on Windows - worst case one can
> create separate code which has the same functionality and interface.
>
Well, this is not necessarily so simple because what I need is timeout for
a tactic -- you timeout on calls to internal Coq API. Apparently,
the Coq/OCaml teams couldn't manage to get this right for quite some time
(see the comments in src/plugin/timeout.ml), but the bugs mentioned there
have been mostly fixed, so maybe now it's possible to use Coq's timeout.

> I would have to have a deeper look into the parallel execution. I guess
> the goal is to spawn several ATP processes in parallel in a controlled way.
> I would think the easiest portable way of doing this would be using the
> OCaml thread library and spawning processes from the threads.
>
Yes, to spawn several ATPs, *but* also to try reconstruction tactics in
parallel (after you get a response from an ATP you try to reconstruct
the proof in Coq with various methods). Spawning ATPs could probably be
done with threads. As far as I remember, though, the problem with OCaml
threads was that they are not truly parallel, i.e., they won't run
physically parallel on several cores. So using them for parallel tactic
invocation which happens inside the plugin makes less sense. Besides,
threads do share memory, and the idea for parallel tactic invocation is
essentially to spawn different Coq processes from a given point to try
different tactics in parallel without messing with each other's internal
Coq state.

**Comment by lukaszcz** [comment]
What I see feasible and I think probably would work:
1. Use Coq's timeout mechanism, which has been fixed.
2. Spawn ATPs with threads.
3. Run reconstruction tactics sequentially.

The third point would substantially reduce CoqHammer usability on
Windows, but not to the point that it couldn't be used. I would really
like to have a solution for parallel tactic invocation on Windows, but
I don't see any right now that doesn't use something like fork.

**Comment by lukaszcz** [comment]
>
> Set Hammer GSMode 0.
> Unset Hammer Parallel.
>

You might also need:

Set Hammer SAutoLimit 0.

This disables the initial call to "sauto" before invoking the provers,
which uses the fork-based timeout.

With those three options you should be able to run provers
sequentially (not 100% sure - haven't tested this legacy mode in quite
a while), but things will probably break when afterwards the proof
reconstruction part tries to use fork.

**Comment by MSoegtropIMC** [comment]
@lukaszcz: a possibly interesting point the `par:` goal selector does work on Windows (or did when I last tried it) - this works by having a pool of coqtops spawned initially and then have a work balancing. I think you might want to reuse this infrastructure - what you want to achieve sounds fairly similar. Another advantage of this is that there are (afaik) mechanisms to globally balance this e.g. in a situation of a parralel make of coq code. Afaik it is so that one can set a global upper limit for the numer of workers spawned globaly. Afaik @gares is the expert for this.

I will try the Coq settings you explained on Windows.

**Comment by MSoegtropIMC** [comment]
Just for my understanding: what you currently do is run coqtop to the point of the tactic application, spawn some ATPs and collect the results. Then you fork coqtop and in each fork you try to convert one of the ATP results into a Gallina term. Is this approximately correct?

**Comment by lukaszcz** [comment]
>
>
> Just for my understanding: what you currently do is run coqtop to the
> point of the tactic application, spawn some ATPs and collect the results.
>
Yes, essentially. Just before spawning the ATPs, an automated tactic inside
the Coq plugin is run with a short timeout to check if it can solve the
goal without invoking anything external (that's what you disable with Set
Hammer SAutoLimit 0).

> Then you fork coqtop and in each fork you try to convert one of the ATP
> results into a Gallina term. Is this approximately correct?
>
Yes, but it's done with different methods for a single ATP result. Once a
single ATP succeeds, we kill the others and start proof reconstruction (the
"conversion"). The "conversion" is just trying various custom tactics to
prove the goal, using the ATP result essentially as hints for the tactics.

**Comment by gares** [comment]
The infrastructure is described correctly, but I would need to look at the code to see if it is easy to pick up. pointers to relevant sources are welcome.

**Comment by lukaszcz** [comment]
@lukaszcz <https://github.com/lukaszcz>: a possibly interesting point the
> par: goal selector does work on Windows (or did when I last tried it) -
> this works by having a pool of coqtops spawned initially and then have a
> work balancing. I think you might want to reuse this infrastructure - what
> you want to achieve sounds fairly similar. Another advantage of this is
> that there are (afaik) mechanisms to globally balance this e.g. in a
> situation of a parralel make of coq code. Afaik it is so that one can set a
> global upper limit for the numer of workers spawned globaly. Afaik @gares
> <https://github.com/gares> is the expert for this.
>
> I will try the Coq settings you explained on Windows.
>
It seems the point of par: is to run the same tactic in parallel on
different subgoals. What I need is running different tactics on one goal
with a timeout and learning which one succeeded. I'm not sure if
this can be achieved with the internal mechanism that's behind "par:".
Maybe @gares <https://github.com/gares> can say.

**Comment by lukaszcz** [comment]
What I need is implementing this file without fork:

https://github.com/lukaszcz/coqhammer/blob/coq8.13/src/lib/hhpartac.ml

**Comment by MSoegtropIMC** [comment]
> t seems the point of par: is to run the same tactic in parallel on
different subgoals.

`par:` is just one use of the infrastructure. The same infrastructure is also used to e.g. prove different lemmas in parallel. Also with `par:` one can duplicate the goal adding say some index hypothesis to it and then run the same tactic on each goal such that the tactic picks up the index and depending on it does different things. I am actually doing such things together with timeout.

**Comment by gares** [comment]
I gave a quick look at the code. the main difference is that you can run any tactic in par mode, while `par:` (and the infrastructure) works fine for "top level tactics". Are you really running your par deep in some ltac code? If it is shallow, then it should be easy to port.

**Comment by lukaszcz** [comment]
No, I think my usage is shallow, i.e., to try various tactics inside 'hammer' and 'best' which are supposed to be called at the "toplevel", i.e., they're "interactive" and it doesn't make much sense to use them somewhere deeper inside ltac code. If that's what you mean.

There are exactly two places where partac is used:
https://github.com/lukaszcz/coqhammer/blob/coq8.13/src/plugin/hammer_main.ml#L596
https://github.com/lukaszcz/coqhammer/blob/coq8.13/src/tactics/tacbest.ml#L456

**Comment by lukaszcz** [comment]
One potential issue is the initial call to "sauto" before invoking the provers. This actually makes the "partac" invocation (which happens after a prover succeeds) wrapped inside Proofview.tclORELSE:
https://github.com/lukaszcz/coqhammer/blob/coq8.13/src/plugin/hammer_main.ml#L773

And I do need access to the the goal to do the translation first, so in any case this must be inside Proofview.Goal.enter or something like this.

**Comment by lukaszcz** [comment]
Summing up, the calls to partac currently are inside the tactic monad, but at a fairly "shallow" level.

**Comment by MSoegtropIMC** [comment]
Adding this to the Coq Platform tracker issue, so that I don't forget to update to the new version.
(https://github.com/coq/platform/issues/139)

**Comment by MSoegtropIMC** [comment]
@lukaszcz : I wanted to ask what the status of Coq 8.14 support is. The Coq Platform for 8.14.0 is expected to be delivered Nov 30 and should include at least coq-hammer-tactics. I would prefer to include coq-hammer as well. I meanwhile did an opam package for eprover - the package for the special variant of Z3 shouldn't be hard.

**Comment by palmskog** [comment]
@lukaszcz if you want to help to set up a new branch for 8.14 with opam packages and the like, please let us know here. Recall that the Coq branching from `master` to `v8.14` happened around June this year.

**Comment by MSoegtropIMC** [comment]
@lukaszcz : I think we can also find support with adjusting coq-hammer to the Coq changes from 8.13.2 to 8.14. It would really help us if you would tell us what your plans / issues are.

**Comment by lukaszcz** [comment]
I'll do it this week. The master branch is being continually kept up to date with the latest Coq development version, so this is usually a matter of just making a branch at an appropriate point, which I normally do when an opam package for Coq is released.

**Comment by MSoegtropIMC** [comment]
@lukaszcz : I just wanted to ask what the status of this is.

I plan to do the special z3 opam package this weekend (e-prover is already done). I would say we publish coq-hammer with these two provers then and add more later. I might need some help with finishing and testing coq-hammer in Coq Platform this week - say a 1 hour online meeting. Would you be available?

**Comment by lukaszcz** [comment]
I will do the 8.14 version today.

Yes, I can do an online meeting. I'll send you a private email later today.

**Comment by lukaszcz** [comment]
@MSoegtropIMC Okay, I don't know how to send you a private message. Sorry, I wasn't available during the weekend up to now. I'm doing the 8.14 release now. If you want to talk with me about something, I have time today or we can arrange something next week. You can find my email on my webpage linked to the github profile.

**Comment by MSoegtropIMC** [comment]
Coq-hammer with eprover and z3_tptp is now included in Coq Platform (Mac and Linux only, Windows contains just the tactics).
